lib/Service/ContactsIntegration.php

```diff
@@ -54,12 +67,28 @@ public function getMatchingRecipient(string $term): array {
 		// If 'Allow username autocompletion in share dialog' is disabled in the admin sharing settings, then we must not
 		// auto-complete system users
 		$allowSystemUsers = $this->config->getAppValue('core', 'shareapi_allow_share_dialog_user_enumeration', 'no') === 'yes';
+		$allowSystemUsersInGroupOnly = $this->config->getAppValue('core', 'shareapi_restrict_user_enumeration_to_group', 'no') === 'yes';
 
 		$result = $this->contactsManager->search($term, ['FN', 'EMAIL']);
 		$receivers = [];
 		foreach ($result as $r) {
-			if (!$allowSystemUsers && isset($r['isLocalSystemBook']) && $r['isLocalSystemBook']) {
-				continue;
+			if (isset($r['isLocalSystemBook']) && $r['isLocalSystemBook']) {
+				if (!$allowSystemUsers) {
+					continue;
+				}
+				if ($allowSystemUsersInGroupOnly) {
+					$userGroups = $this->groupManager->getUserGroupIds($this->userSession->getUser());
```

Comment:
Nitpick: the user should be an input argument of the method.

Background: that makes the method independent of the context. E.g. some processes run in the background. They will then still work and not produce unpredictable results.